src/adapter.js

```diff
@@ -103,7 +125,47 @@ class MongooseAdapter {
   }
 
   setTransaction (transactioned = true) {
-    this.useTransaction = transactioned;
+    const conn = this.mongoseInstance.connections[0];
+    if (!transactioned) this.useTransaction = transactioned;
+    else if (conn && conn.replica) this.useTransaction = transactioned;
+    else throw Error('Tried to enable transactions for non-replicaset connection');
+  }
+
+  async getSession () {
+    if (this.useTransaction) {
+      this.session = this.session && !this.session.hasEnded() ? this.session : await CasbinRule.startSession();
+      return this.session;
+    } else throw Error('Tried to start a session for non-replicaset connection');
+  }
+
+  async getTransaction () {
+    if (this.useTransaction) {
+      try {
+        const session = await this.getSession();
+        await session.startTransaction();
+        return session;
+      } catch (error) {
+        if (!error.message !== 'Transaction already in progress on this session.') throw error;
```

Comment:
What will happen if some other MongoDB version will not have this error message? You will not be able to detect "transaction in progress" at all and make a false negative.

Maybe MongoDB has some codes for it, take a look into documentation

Reply:
Yeah, took a look into the documentation while trying to figure out how to integrate replica / transactions for the adapter. I guess I saw some types for the errors at one point, so that could be a better way to do it.

Reply:
https://github.com/mongodb/mongo/blob/master/src/mongo/shell/session.js#L674

It just throws a normal error without anything else. Looking through the source code and trying with the mongo shell, tnx state variables are not accessible.

This might be the best we can do. However the Error message has been around for a few years.

Reply:
Actually, TnxState can be accessed via:
```node
session._serverSession.handle.getTxnState()
```
Which varies between active, inactive etc. This could be more solid way to do it, but I do not really see anybody else doing this.

Reply:
Yeah, seems like undocumented feature, is it?

I remember mongoose errors have some codes in `Error`, no?

Reply:
As it's the same error linked above (and it comes straight back without any other middleware), there won't be any other information about the error, other than the stack trace.

`_serverSession` is definitely meant as "private" object, but there is no other way to reliably get access to more meaningful status attributes of the session / transaction. Transactions in general are more advanced feature since it's only usable by replicas, which are little bit more under used features of mongoDB (and more generally hosted / managed by the MongoDB themselves).

Documentation and API implementation of sessions and transactions does not seem as finely tuned as the rest of mongoDB features. It seems to cover just the basics.

I think the current solution in the PR is quite good implementation so far (it's the same that mongo itself uses).

Reply:
If we don't know any other alternatives at the moment, so it will be good enough. But I'm afraid there is a room for false negatives :(